### PR TITLE
Introduce deductive annotations

### PR DESCRIPTION
**Pull request by david-yz-liu**
## Motivation and Context

One of the ways digital marking tools can be leveraged optimally is by speeding up the process of highlighting problems with submissions, while simultaneously reducing the given grade by the appropriate amount. Reusable commenting on submissions already exists in MarkUs as AnnotationTexts that belong to AnnotationCategories. What would greatly aid in the speed and simplicity of marking with these reusable annotations, is if they were somehow linked to a grade reduction, and would automatically apply that grade reduction when used on a submission. This kind of functionality is what this pull request aims to address, and allows graders to reuse comments, with associated penalties, that may be applicable to many submissions.

## Your Changes

**Description**:
This pull request is the culmination of many pull requests on the `deductive-annotations` branch. Essentially, the pull request includes but is not limited to the following:

- A migration of the database
- Model changes and additional callbacks
- Additional routes
- Changes to the views
- Changes to the seed data
- Additions to rspec tests and factories

**Type of change** (select all that apply):

 
- [x] New feature (non-breaking change which adds functionality)

## Testing

Each pull request that was made to this branch in the process of its creation was vetted by Hound, TravisCI, and Coveralls. Additionally, test examples were added as changes were made to the server side code, and the changes to the views were verified to work as expected in both Firefox & Chrome.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have added tests for my changes, if applicable.
- [ ] I have updated the Changelog.md file.
- [x] I have described any required documentation changes below, if applicable.
- [ ] I have fixed any Hound bot comments (check after opening pull request).
- [ ] I have verified that the TravisCI tests have passed (check after opening pull request).
- [ ] I have reviewed the test coverage changes reported on Coveralls (check after opening pull request).

### Required documentation changes (if applicable)
Changes were made to the MarkUs Wiki on GitHub. These included two new pages ([Instructor Guide: Setting Up Deductive Annotations](https://github.com/MarkUsProject/Markus/wiki/Instructor-Guide:-Setting-Up-Deductive-Annotations), [Instructor Guide: Marking with Deductive Annotations](https://github.com/MarkUsProject/Markus/wiki/Instructor-Guide:-Marking-with-Deductive-Annotations)) as well as changes to the existing page [Importing and Exporting Data](https://github.com/MarkUsProject/Markus/wiki/Importing-and-Exporting-Data).

### CODE REVIEW

File: app/controllers/annotation_categories_controller.rb [hunk 1]
```diff
@@ -71,6 +75,7 @@ def update
 
     if @annotation_category.update(annotation_category_params)
       flash_message(:success, t('.success'))
+      render 'show', assignment_id: @assignment.id, id: @annotation_category.id
```
Comment (ignaspa) [inline app/controllers/annotation_categories_controller.rb:78]: This is because the refreshing of the UI should happen in cases where the criterion is updated.

File: app/models/mark.rb [hunk 2]
```diff
@@ -18,6 +19,44 @@ class Mark < ApplicationRecord
   validates_uniqueness_of :markable_id,
                           scope: [:result_id, :markable_type]
 
+  validates_inclusion_of :override, in: [true, false]
+
+  def calculate_deduction
+    return 0 if self.markable_type != 'FlexibleCriterion' || self.override?
```
Comment (ignaspa) [inline app/models/mark.rb:25]: Possibly remove `|| self.override` as long as every place calling this method checks that.

File: config/locales/views/annotation_categories/en.yml [hunk 3]
```diff
@@ -4,7 +4,12 @@ en:
       success: "This annotation category was successfully created."
     destroy_annotation_text:
       confirm: "Deleting this annotation will remove the text from any files it may be attached to.  Are you sure you want to do this?"
+      error: 'Cannot delete this annotation'
```
Comment (ignaspa) [inline config/locales/views/annotation_categories/en.yml:7]: Perhaps give a more explicit reason (same for `update_annotation_text` below) Like "Cannot delete this annotation because it has a deductive value greater than 0 and has been applied to released results."

File: app/models/annotation_category.rb [hunk 4]
```diff
@@ -15,15 +30,86 @@ def self.add_by_row(row, assignment, current_user)
...
-        last_editor_id: current_user.id
-      )
-      unless annotation_text.save
-        raise CsvInvalidLineError
+  def marks_released?
+    !self.assignment.released_marks.empty?
+  end
+
+  def deductive_annotations_exist?
+    !self.annotation_texts.joins(:annotations).where.not(deduction: nil).empty?
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/models/annotation_category.rb:83]: use exists instead of empty, remove `!`

File: config/locales/views/annotation_categories/en.yml (hunk 3 above)
Comment (ignaspa) [reply config/locales/views/annotation_categories/en.yml:7]: Change key as well, deductive_annotation_released_error.

File: app/models/annotation_text.rb [hunk 5]
```diff
@@ -3,15 +3,58 @@ class AnnotationText < ApplicationRecord
...
+  end
+
+  def update_mark_deductions
+    return unless previous_changes.key?('deduction')
+
+    # depending on category to change marks if deduction becomes nil
+    return if self.deduction.nil?
+
+    annotations = self.annotations.includes(:result)
+    annotations.each do |annotation|
```
Comment (ignaspa) [inline app/models/annotation_text.rb:48]: Use a `through` association for results of an annotation text.

File: app/controllers/results_controller.rb [hunk 6]
```diff
@@ -561,6 +586,25 @@ def update_mark
     end
   end
 
+  def revert_to_automatic_deductions
+    result = Result.find(params[:id])
+    result_mark = result.marks.find_or_create_by(markable_id: params[:markable_id], markable_type: 'FlexibleCriterion')
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/controllers/results_controller.rb:591]: Leaving this as is because the `mark` creation lifecycle is nebulous.

File: app/models/annotation_text.rb [hunk 7]
```diff
@@ -3,15 +3,58 @@ class AnnotationText < ApplicationRecord
...
+
   def escape_content
     content.gsub('\\', '\\\\\\') # Replaces '\\' with '\\\\'
            .gsub(/\r?\n/, '\\n')
   end
+
+  def check_if_released
+    # Cannot update if any results have been released with the annotation and the deduction is non nil
+    return if self.annotations.joins(:result).where('results.released_to_students' => true).empty? ||
+        self.deduction.nil?
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/models/annotation_text.rb:36]: We will keep this as limited as it is for now, because the development of the platform is hopefully moving in the direction of this functionality holding true for all annotations.

File: app/models/mark.rb (hunk 2 above)
Comment (ignaspa) [reply app/models/mark.rb:25]: This helper method should stand on its own for the relevant conditions and not be dependent on existing calls or calls implemented in the future to check this condition.

File: spec/controllers/annotations_controller_spec.rb [hunk 8]
```diff
@@ -239,11 +256,133 @@
...
+                :destroy,
+                params: { id: annotation.id,
+                          result_id: result.id,
+                          assignment_id: assignment.id },
+                format: :js
+        assert_response :bad_request
+        expect(result.reload.annotations.size).to eq 1
+      end
+
+      it 'cannot destroy a deductive annotation if assigned to a criterion of a different type '\
```
Comment (ignaspa) [inline spec/controllers/annotations_controller_spec.rb:347]: I was thinking:
Similar to this test, but perhaps missing here, is a test to check if the TA cannot destroy a deductive annotation if assigned to a criterion of the same type and id as one for deductive annotations in an assignment, but for a different assignment. 

But I realized:
The same criterion type can never have the same ID regardless of assignment.

However:
Should there be a test for criteria across assignments?

File: spec/controllers/results_controller_spec.rb [hunk 9]
```diff
@@ -322,6 +369,70 @@ def self.test_unauthorized(route_name)
...
+          assignment_id: assignment.id,
+          submission_id: assignment.groupings.first
+                                   .current_result
+                                   .submission.id,
+          id: assignment.groupings.first.current_result,
+          markable_id: mark.markable_id,
+          format: :json
+        }, xhr: true
+
+        expect(response.parsed_body.key?('sub_total')).to be true
```
Comment (ignaspa) [inline spec/controllers/results_controller_spec.rb:430]: Same thing with the keys here.

File: spec/controllers/results_controller_spec.rb [hunk 10]
```diff
@@ -540,6 +708,22 @@ def self.test_unauthorized(route_name)
       end
     end
 
+    context 'when criteria are assigned to graders, but not this grader' do
+      it 'receives no deductive annotation category data if has no ta_criterion_associations' do
```
Comment (ignaspa) [inline spec/controllers/results_controller_spec.rb:712]: remove description after 'data'. It is implied by the context.

File: spec/controllers/results_controller_spec.rb [hunk 11]
```diff
@@ -553,6 +737,46 @@ def self.test_unauthorized(route_name)
         expect(data['assigned_criteria']).to eq ["#{rubric_criterion.class}-#{rubric_criterion.id}"]
       end
 
+      context 'when accessing an assignment with deductive annotations' do
+        let(:assignment) { create(:assignment_with_deductive_annotations) }
+        it 'receives limited annotation category data if has ta_criterion_association' do
```
Comment (ignaspa) [inline spec/controllers/results_controller_spec.rb:742]: Change 'if has ta_criterion_association' to 'when assigned to a subset of criteria that have associated categories'

File: spec/controllers/results_controller_spec.rb [hunk 12]
```diff
@@ -553,6 +737,46 @@ def self.test_unauthorized(route_name)
...
+                                submission_id: assignment.groupings.first.current_result.submission.id,
+                                id: assignment.groupings.first.current_result,
+                                format: :json }, xhr: true
+          expect(response.parsed_body['annotation_categories'].first['annotation_category_name'])
+            .to eq "#{other_category.annotation_category_name} [#{other_category.flexible_criterion.name}]"
+          expect(response.parsed_body['annotation_categories'].size).to eq 1
+        end
+
+        it 'receives limited annotation category data even when it has a ta_criterion_association '\
+           'with a criterion that has the same id as a flexible criterion' do
```
Comment (ignaspa) [inline spec/controllers/results_controller_spec.rb:762]: continue ...'which is associated with a category'

File: spec/models/annotation_category_spec.rb [hunk 13]
```diff
@@ -75,4 +124,72 @@
...
+      expect { assignment.annotation_categories.destroy_all }.to raise_error ActiveRecord::RecordNotDestroyed
+    end
+
+    it 'does not prevent deletion of an annotation_category if annotations have '\
+       'no deduction and results not released' do
+      annotation_category_with_criteria.update!(flexible_criterion_id: nil)
+      expect { assignment.annotation_categories.destroy_all }.to_not raise_error
+    end
+
+    it 'does not prevent deletion of an annotation_category if annotations have no deduction and results released' do
```
Comment (ignaspa) [inline spec/models/annotation_category_spec.rb:185]: Add example for if all deductions are 0. (should allow deletion)

File: spec/models/annotation_category_spec.rb [hunk 14]
```diff
@@ -34,6 +81,7 @@
       before do
         @row = []
         @row.push('annotation category name 2')
+        @row.push(nil)
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply spec/models/annotation_category_spec.rb:84]: Raise error if upload of annotation category with same name as existing annotation category changes the criterion id. Apply to all scenarios. (three tests for possible id changes)

[the remark this answers is not in the document]
Comment (ignaspa) [reply spec/models/annotation_category_spec.rb:84]: Because the annotation text creation in case of either criterion being specified or the case where the criterion is not specified is a call to `build`, we know that the annotation texts will never be updated during an upload. They will only be created. (to address point 3)

File: spec/models/annotation_text_spec.rb [hunk 15]
```diff
@@ -24,4 +24,135 @@
...
+    end
+  end
+
+  describe '#update_mark_deductions' do
+    it 'updates the mark associated with its annotation category\'s flexible criterion' \
+       ' for every grouping if its deduction changed' do
+      deductive_text.update!(deduction: 2.0)
+      assignment.reload
+      marks = []
+      assignment.groupings.includes(:current_result).each do |grouping|
```
Comment (ignaspa) [inline spec/models/annotation_text_spec.rb:99]: Simplify to `joins + where + pluck`, similar throughout.

File: app/models/annotation_text.rb (hunk 5 above)
Comment (ignaspa) [reply app/models/annotation_text.rb:48]: `through` associations will be implemented as possible through patch PRs to simplify code.

File: app/assets/javascripts/Components/Result/annotation_table.jsx [hunk 16]
```diff
@@ -99,6 +104,50 @@ export class AnnotationTable extends React.Component {
...
+    id: 'deduction',
+    filterMethod: this.deductionFilter,
+    Cell: data => {
+      if (data.original.deduction === undefined || data.original.deduction === null ||
+          data.original.deduction === 0.0) {
+        return '';
+      } else {
+        return (
+          <div>
+            <span>
```
Comment (ignaspa) [inline app/assets/javascripts/Components/Result/annotation_table.jsx:126]: This span might be unnecessary. (check, then remove if so)

File: app/assets/javascripts/Components/Result/marks_panel.jsx [hunk 17]
```diff
@@ -229,6 +232,68 @@ class FlexibleCriterionInput extends React.Component {
...
 
+  listDeductions = () => {
+    let label = I18n.t('annotations.list_deductions');
+    let deductiveAnnotations = this.props.annotations.filter(a => {
+      return a.criterion_id !== undefined && a.criterion_id !== null &&
+        a.deduction !== 0.0 && a.criterion_id === this.props.id;
+    });
+
+    if (deductiveAnnotations.length === 0) {
+      return <span></span>;
```
Comment (ignaspa) [inline app/assets/javascripts/Components/Result/marks_panel.jsx:243]: Can this just be an empty string so there isn't an element?

File: app/assets/javascripts/Components/Result/marks_panel.jsx [hunk 18]
```diff
@@ -229,6 +232,68 @@ class FlexibleCriterionInput extends React.Component {
...
+        </span>
+        <span className={'red-text'}>
+          {hyperlinkedDeductions}
+        </span>
+      </div>);
+  }
+
+  deleteManualMarkLink = () => {
+    if (!this.props.released_to_students && !this.props.unassigned) {
+      if (this.props.annotations.some(a => a.deduction !== null && a.deduction !== 0 && a.criterion_id ===
```
Comment (ignaspa) [inline app/assets/javascripts/Components/Result/marks_panel.jsx:279]: Add if deduction is undefined condition.

File: app/assets/javascripts/Components/Result/result.jsx [hunk 19]
```diff
@@ -141,8 +141,12 @@ class Result extends React.Component {
         }];
       } else {
         children = annotation_category.texts.map(text => {
+          let deduction = '';
+          if (text.deduction !== undefined && text.deduction !== null && text.deduction !== 0) {
+            deduction = '-' + text.deduction;
+          }
           return {
-            title: text.content.replace(/\r?\n/gi, ' '),
+            title: `${text.content.replace(/\r?\n/gi, ' ')} <span class="red-text">${deduction}</span>`,
```
Comment (ignaspa) [inline app/assets/javascripts/Components/Result/result.jsx:149]: Perhaps add `text-content` class to title text so deduction is to the right of the element.

File: app/assets/stylesheets/common/_markus.scss [hunk 20]
```diff
@@ -647,6 +647,7 @@ ul.tags {
       li {
         color: $dark-grey;
         cursor: pointer;
+        display: inline-flex;
```
Comment (ignaspa) [inline app/assets/stylesheets/common/_markus.scss:650]: Not sure what this is for.

File: app/views/annotation_categories/_annotation_category_edit.html.erb [hunk 21]
```diff
@@ -19,6 +20,16 @@
         <%= f.label :annotation_category_name,
                     AnnotationCategory.human_attribute_name(:annotation_category_name) %>
         <%= f.text_field :annotation_category_name, rows: '3' %>
+        <%= f.label :flexible_criterion_id,
+                    AnnotationCategory.human_attribute_name(:flexible_criterion) %>
+        <%= f.collection_select :flexible_criterion_id,
```
Comment (ignaspa) [inline app/views/annotation_categories/_annotation_category_edit.html.erb:25]: Why ta_visible? Also, possible repetition of assignment specification for the criterion.

File: app/views/annotations/create.js.erb [hunk 22]
```diff
@@ -1,29 +1,52 @@
...
+    id: <%= @annotation.id %>,
+    creator: '<%= "#{@annotation.creator.first_name} #{@annotation.creator.last_name}" %>',
+    filename: '<%= @annotation.submission_file.filename %>',
+    path: '<%= @annotation.submission_file.path.split('/', 2)[1] %>',
+    submission_file_id: <%= @annotation.submission_file_id %>,
+    content: '<%= @annotation.annotation_text.escape_content %>',
+    annotation_category: '<%= @annotation.annotation_text.annotation_category&.annotation_category_name %>',
+    annotation_text_id: <%= @annotation.annotation_text_id %>,
+    criterion_id: <%= @annotation.annotation_text.annotation_category&.flexible_criterion_id || 'undefined' %>,
+    criterion_name: '<%= @annotation.annotation_text.annotation_category&.flexible_criterion&.name || 'undefined' %>',
```
Comment (ignaspa) [inline app/views/annotations/create.js.erb:12]: Check one last time, is this correct notation?

File: spec/models/annotation_category_spec.rb [hunk 23]
```diff
@@ -9,16 +9,63 @@
...
+
+    it 'returns an error message if a criterion with the name given does not exist' do
+      row = ['category_name', 'criterion_name', 'text_content', '1.0']
+      expected_message = I18n.t('annotation_categories.upload.criterion_not_found',
+                                missing_criterion: 'criterion_name')
+      expect { AnnotationCategory.add_by_row(row, assignment, admin) }.to raise_error(CsvInvalidLineError,
+                                                                                      expected_message)
+    end
+
+    it 'returns an error message if a deduction given for an annotation text '\
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply spec/models/annotation_category_spec.rb:53]: Actually, just automatically round to two decimal points.

File: app/views/annotation_categories/_annotation_category_edit.html.erb (hunk 21 above)
Comment (ignaspa) [reply app/views/annotation_categories/_annotation_category_edit.html.erb:25]: We can be less restrictive, remove this specificity.

File: app/views/annotations/create.js.erb (hunk 22 above)
Comment (ignaspa) [reply app/views/annotations/create.js.erb:12]: No, should be ...`ory&.flexible_criterion&.name %>' || undefined,`

File: app/assets/stylesheets/common/_markus.scss (hunk 20 above)
Comment (ignaspa) [reply app/assets/stylesheets/common/_markus.scss:650]: So child elements respect flex formatting.

File: spec/models/mark_spec.rb [hunk 24]
```diff
@@ -70,6 +73,132 @@
...
+      3.times do
+        create(:text_annotation,
+               annotation_text: annotation_text,
+               result: result)
+      end
+      result.reload
+      expect(mark.mark).to eq(0.0)
+    end
+
+    it 'does not change the mark when there are no deductions' do
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply spec/models/mark_spec.rb:148]: This would mean a situation where the mark is non-nil and there are no deductive annotations, which means `override` must be enabled. There is already an example for this.

File: app/assets/javascripts/Components/Result/result.jsx (hunk 19 above)
Comment (ignaspa) [reply app/assets/javascripts/Components/Result/result.jsx:149]: No effect, and hard to debug due to content menu disappearing when inspecting element.

File: spec/controllers/annotation_categories_controller_spec.rb [hunk 25]
```diff
@@ -33,6 +34,16 @@
 
       expect(assignment.annotation_categories.count).to eq 1
     end
+
+    it 'successfully creates annotation_category with nil flexible_criterion' do
+      category = create(:annotation_category, assignment: assignment)
```
Comment (david-yz-liu) [inline spec/controllers/annotation_categories_controller_spec.rb:39]: I don't think you want to do this here. Category names need to be unique for an assignment, so I suspect the `post` call below is not actually creating a valid object.

Instead, you can remove this and pass in a hard-coded string (e.g., `'Category 1'`) to the `post`.

File: spec/controllers/annotation_categories_controller_spec.rb [hunk 26]
```diff
@@ -235,6 +387,29 @@
       expect(found_cat).to eq(true)
     end
 
+    it 'accepts a valid csv file with deductive annotation info' do
+      file_good = fixture_file_upload('files/annotation_categories/form_good_with_deductive_info.csv',
+                                      'text/csv')
+      create(:flexible_criterion, name: 'hephaestus', assignment: assignment)
+      post :upload, params: { assignment_id: assignment.id, upload_file: file_good }
+
+      expect(response.status).to eq(302)
```
Comment (david-yz-liu) [inline spec/controllers/annotation_categories_controller_spec.rb:396]: Here and below, don't put parentheses around the argument to `eq`

File: spec/controllers/results_controller_spec.rb [hunk 27]
```diff
@@ -505,6 +613,62 @@ def self.test_unauthorized(route_name)
...
+                                   flexible_criterion_id: other_criterion.id)
+          second_name = "#{second_category.annotation_category_name} [#{second_category.flexible_criterion.name}]"
+          post :show, params: { assignment_id: assignment.id,
+                                submission_id: assignment.groupings.first.current_result.submission.id,
+                                id: assignment.groupings.first.current_result,
+                                format: :json }, xhr: true
+
+          category_names = [first_name, second_name].sort!
+          returned_categories = []
+          response.parsed_body['annotation_categories'].each do |cat|
```
Comment (david-yz-liu) [inline spec/controllers/results_controller_spec.rb:641]: Use `.map` here.

File: spec/models/annotation_category_spec.rb [hunk 28]
```diff
@@ -9,16 +9,99 @@
...
+      let(:category) { assignment.annotation_categories.where.not(flexible_criterion_id: nil).first }
+
+      it 'does not allow the flexible_criterion_id to be set to reference a criterion of another assignment' do
+        other_assignment = create(:assignment_with_criteria_and_results)
+        other_criterion_id = other_assignment.flexible_criteria.first.id
+        category.flexible_criterion_id = other_criterion_id
+        expect(category).to_not be_valid
+      end
+
+      it 'allows the flexible_criterion_id to be set to reference a criterion of it\'s assignment' do
```
Comment (david-yz-liu) [inline spec/models/annotation_category_spec.rb:29]: `it's` > `its`

File: spec/models/annotation_category_spec.rb [hunk 29]
```diff
@@ -9,16 +9,99 @@
...
+        other_criterion = create(:flexible_criterion, assignment: assignment)
+        category.flexible_criterion_id = other_criterion.id
+        expect(category).to be_valid
+      end
+    end
   end
 
   describe '.add_by_row' do
+    it 'returns an error message if the category name is blank' do
+      row = [nil, 'criterion_name', 'text_content', '1.0']
```
Comment (david-yz-liu) [inline spec/models/annotation_category_spec.rb:39]: It''s a more realistic test to use `''` instead of `nil` here.

File: spec/models/annotation_category_spec.rb [hunk 30]
```diff
@@ -45,12 +129,44 @@
         AnnotationCategory.add_by_row(@row, assignment, admin)
         expect(@initial_size + 1).to eq(AnnotationCategory.all.size)
       end
+
+      context 'and the assignment has deductive annotations' do
+        let(:assignment) { create(:assignment_with_deductive_annotations) }
+        let(:category) { assignment.annotation_categories.where.not(flexible_criterion_id: nil).first }
+
+        it 'does not allow the flexible criterion to changed to nil' do
```
Comment (david-yz-liu) [inline spec/models/annotation_category_spec.rb:137]: changed -> change (and below as well)

File: spec/models/annotation_category_spec.rb [hunk 31]
```diff
@@ -45,12 +129,44 @@
...
+        it 'does not allow the flexible criterion for the category to changed to a different criterion' do
+          other_criterion = create(:flexible_criterion, assignment: assignment)
+          row = [category.annotation_category_name, other_criterion.name, 'text 1', 1.0]
+          expected_message = I18n.t('annotation_categories.upload.invalid_criterion',
+                                    annotation_category: category.annotation_category_name)
+          expect { AnnotationCategory.add_by_row(row, assignment, admin) }.to raise_error(CsvInvalidLineError,
+                                                                                          expected_message)
+        end
+
+        it 'does not allow the criterion for the category to be specified it was not before' do
```
Comment (david-yz-liu) [inline spec/models/annotation_category_spec.rb:154]: I think this description is missing a word or two.

File: spec/models/annotation_spec.rb [hunk 32]
```diff
@@ -21,4 +21,50 @@
...
+    let(:annotation_category) { assignment.annotation_categories.where.not(flexible_criterion_id: nil).first }
+    let(:annotation_text) { annotation_category.annotation_texts.first }
+    let(:result) { assignment.groupings.first.current_result }
+    let(:mark) { result.marks.find_by(markable_id: annotation_category.flexible_criterion_id) }
+
+    it 'correctly updates the mark when created' do
+      create(:text_annotation,
+             annotation_text: annotation_text,
+             result: result)
+      expect(mark.mark).to eq(1.0)
```
Comment (david-yz-liu) [inline spec/models/annotation_spec.rb:36]: Remove parentheses around argument to `eq` (and below)

File: spec/models/annotation_text_spec.rb [hunk 33]
```diff
@@ -24,4 +24,135 @@
...
+    it 'is invalid with a numerical deduction if it does not belong to flexible criteria through annotation category' do
+      text_without_deduction.deduction = 1.0
+      expect(text_without_deduction).to_not be_valid
+    end
+  end
+
+  describe 'callbacks' do
+    it 'prevent an update of deduction if any results are released' do
+      assignment.groupings.first.current_result.update!(released_to_students: true)
+      expect(deductive_text.update(deduction: 2.0)).to be(false)
```
Comment (david-yz-liu) [inline spec/models/annotation_text_spec.rb:71]: Similar to above, remove parentheses around argument to `be`

File: spec/models/annotation_text_spec.rb (hunk 15 above)
Comment (david-yz-liu) [inline spec/models/annotation_text_spec.rb:99]: Use `map` here (and below).

File: spec/models/annotation_text_spec.rb [hunk 34]
```diff
@@ -24,4 +24,135 @@
...
+  let(:assignment) { create(:assignment_with_deductive_annotations) }
+  let(:annotation_category_with_criterion) do
+    assignment.annotation_categories.where.not(flexible_criterion_id: nil).first
+  end
+  let(:deductive_text) do
+    annotation_category_with_criterion.annotation_texts.where.not(deduction: nil).first
+  end
+  let(:category_without_criteria) { create(:annotation_category, assignment: assignment) }
+  let(:text_without_deduction) { create(:annotation_text, annotation_category: category_without_criteria) }
+
```
Comment (david-yz-liu) [inline spec/models/annotation_text_spec.rb:37]: Add a variable for `:flexible_criterion` and use it below (so you don't have to keep doing `annotation_category_with_criterion.flexible_criterion`).

File: spec/models/annotation_text_spec.rb [hunk 35]
```diff
@@ -24,4 +24,135 @@
...
+  describe '#update_mark_deductions' do
+    it 'updates the mark associated with its annotation category\'s flexible criterion' \
+       ' for every grouping if its deduction changed' do
+      deductive_text.update!(deduction: 2.0)
+      assignment.reload
+      marks = []
+      assignment.groupings.includes(:current_result).each do |grouping|
+        marks << grouping.current_result.marks
+                         .find_by(markable_id: annotation_category_with_criterion.flexible_criterion_id).mark
```
Comment (david-yz-liu) [inline spec/models/annotation_text_spec.rb:101]: Since the markable association is polymorphic, it's better to do `find_by(markable: annotation_category_with_criterion.flexible_criterion)` (avoids any possibility of duplicate ids).

File: spec/models/flexible_criterion_spec.rb [hunk 36]
```diff
@@ -53,6 +53,55 @@
       @assignment = create(:assignment)
     end
 
+    context 'with deductive annotations' do
+      let(:assignment) { create(:assignment_with_deductive_annotations) }
+      let(:flexible_criterion) { assignment.flexible_criteria.first }
+      let(:annotation_category) do
+        assignment.annotation_categories.where(flexible_criterion_id: flexible_criterion.id).first
+      end
+      it 'reassigns it\'s annotation_category\'s flexible_criterion_id to nil before being destroyed if it has one' do
```
Comment (david-yz-liu) [inline spec/models/flexible_criterion_spec.rb:62]: it's -> its

Comment (david-yz-liu) [reply spec/models/flexible_criterion_spec.rb:62]: (and below)

File: spec/models/flexible_criterion_spec.rb [hunk 37]
```diff
@@ -53,6 +53,55 @@
...
+
+      it 'reassigns it\'s annotation_categories\' flexible_criterion_ids to nil before ' \
+         'being destroyed if it has many' do
+        create(:annotation_category,
+               flexible_criterion_id: flexible_criterion.id,
+               assignment: assignment)
+        flexible_criterion.destroy
+        assignment.reload
+        category_criteria = []
+        assignment.annotation_categories.each do |category|
```
Comment (david-yz-liu) [inline spec/models/flexible_criterion_spec.rb:76]: Use `map` here (and below)

File: spec/models/mark_spec.rb [hunk 38]
```diff
@@ -70,6 +73,166 @@
...
+      create(:text_annotation,
+             annotation_text: annotation_text,
+             result: result)
+      result.reload
+      expect(mark.mark).to eq(3.0)
+    end
+
+    it 'does not allow deductions to reduce the mark past 0' do
+      3.times do
+        create(:text_annotation,
```
Comment (david-yz-liu) [inline spec/models/mark_spec.rb:152]: For duplicated object creation, use [`create_list`](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#building-or-creating-multiple-records).

File: lib/tasks/rubric.rake [hunk 39]
```diff
@@ -54,6 +54,26 @@ namespace :db do
             assigned_groups_count:   nil
         )
       end
+      ac_with_criterion = AnnotationCategory.create(assignment: assignment,
+                                                    position: 6,
+                                                    annotation_category_name: random_words(1),
+                                                    flexible_criterion_id: assignment.flexible_criteria.first.id)
+      rand(3..12).times do |index|
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply lib/tasks/rubric.rake:61]: Yeah, I understand. I think I didn't realize this was a style issue, because I transitioned from actually using the index for the annotation texts so it's vestigial.

File: config/locales/views/annotation_categories/en.yml [hunk 40]
```diff
@@ -17,7 +22,12 @@ en:
     update:
       success: "This annotation category was successfully updated."
     upload:
-      error: "Unable to upload category: %{annotation_category} located in line %{annotation_line}"
+      criterion_not_found: "Criterion %{missing_criterion} must exist prior to the upload of the annotation category belonging to it"
+      deduction_absent: 'Deduction values for category %{annotation_category} absent, despite criterion specification.'
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply config/locales/views/annotation_categories/en.yml:26]: So should this be simplified to combine with deduction_invalid key?

File: spec/models/annotation_category_spec.rb (hunk 29 above)
Comment (ignaspa) [reply spec/models/annotation_category_spec.rb:39]: Hm, this is a little complicated, we can discuss.
